File: lib/utils/prophecy_prompt_builder.dart
import '../models/sensor_data.dart';

/// 唯一 prompt 源：统一风格 spec + few-shot + 传感器上下文
class ProphecyPromptBuilder {
  ProphecyPromptBuilder._();

  static const _styleSpec = '''
你是废话预言家。根据传感器数据写一条中文预言。

风格要求：
- 32–42 字，硬上限 45 字
- 第二人称「你」，允许句首使用
- 冷幽默 + 荒诞因果 + 伪科学精度，不要哲理/禅意腔
- 结构：[传感器锚定]，[荒诞预测]
- 禁止前言、自述、emoji、英文
- 鼓励引用传感器字段并含 1 个伪精度数字
- 只输出预言正文，不要引号或前缀''';

  static const _fewShotExamples = [
    '电量72%时，你的拇指滑屏速度会比平时快1.2倍',
    '今日步数3500步，你的手指比预期早了0.3秒划到下一张图',
    '系统音量45%时，你听到的下一句废话会比上一句响0.3分贝',
  ];

  static String _ambientLightLine(SensorData sensor) {
    if (sensor.isRealAmbientLight) {
      return '- 环境光线：${sensor.ambientLight} 勒克斯';
    }
    if (sensor.isEstimatedAmbientLight) {
      return '- 环境亮度（屏幕推算）：约 ${sensor.ambientLight} 勒克斯';
    }
    return '- 环境光线：未知';
  }

  static String _avoidRepeatBlock(String? avoidText, int? nonce) {
    final parts = <String>[];
    if (avoidText != null && avoidText.isNotEmpty) {
      parts.add('- 上一句预言：$avoidText');
      parts.add('- 必须写完全不同的新句子，禁止重复或改写上一句');
    }
    if (nonce != null) {
      parts.add('- 本次编号：$nonce（每次编号不同，内容也必须不同）');
    }
    if (parts.isEmpty) return '';
    return '\n${parts.join('\n')}';
  }

  /// LocalAi 用纯文本 prompt
  static String buildPrompt(
    SensorData sensor, {
    String? avoidText,
    int? nonce,
  }) {
    return '''$_styleSpec

示例：
1. ${_fewShotExamples[0]}
2. ${_fewShotExamples[1]}
3. ${_fewShotExamples[2]}

当前传感器：
- 时段：${sensor.dayPhase} · ${sensor.timeHint}
- 电量：${sensor.battery ?? 50}%
- 屏幕亮度：${sensor.brightness}%
- 系统音量：${sensor.volume}%
- 今日步数：${sensor.steps} 步
- 身体状态：${sensor.isMoving ? '正在移动' : '静止'}
${_ambientLightLine(sensor)}${_avoidRepeatBlock(avoidText, nonce)}

预言：''';
  }

  /// MLX 用 ChatML 包装
  static String buildChatMLPrompt(
    SensorData sensor, {
    String? avoidText,
    int? nonce,
  }) {
    final userContent = '''根据以下传感器数据写一条预言：

- 时段：${sensor.dayPhase} · ${sensor.timeHint}
- 电量：${sensor.battery ?? 50}%
- 屏幕亮度：${sensor.brightness}%
- 系统音量：${sensor.volume}%
- 今日步数：${sensor.steps} 步
- 身体状态：${sensor.isMoving ? '正在移动' : '静止'}
${_ambientLightLine(sensor)}${_avoidRepeatBlock(avoidText, nonce)}

示例：
1. ${_fewShotExamples[0]}
2. ${_fewShotExamples[1]}
3. ${_fewShotExamples[2]}''';

    return '''<|im_start|>system
$_styleSpec
<|im_end|>
<|im_start|>user
$userContent
<|im_end|>
<|im_start|>assistant
''';
  }
}

File: server/app/services/prophecy_prompt_builder.py
"""预言 prompt 构建（移植自 lib/utils/prophecy_prompt_builder.dart）。"""

from __future__ import annotations

from dataclasses import dataclass

from app.models.sensor import SensorPayload

_FEW_SHOT_EXAMPLES = [
    "电量72%时，你的拇指滑屏速度会比平时快1.2倍",
    "你接下来的三分钟内会突然想起一件无关紧要的小事",
    "今天你会在电梯里和陌生人交换一个意味深长的眼神",
]

_SYSTEM_PROMPT = """你是废话预言家。写一句中文废话预言。
要求：32-42字；第二人称「你」；冷幽默荒诞；可不提及传感器数据。
禁止：解释、题面复述、出现「上一句」「传感器」「当前状态」「写一条」等词。
只输出预言正文一行。"""


@dataclass(frozen=True)
class MlxChatPrompt:
    system: str
    user: str


def _sensor_facts(sensor: SensorPayload) -> str:
    battery = sensor.battery if sensor.battery is not None else 50
    motion = "移动" if sensor.is_moving else "静止"
    light = ""
    if sensor.is_real_ambient_light or sensor.is_estimated_ambient_light:
        light = f"，光线{sensor.ambient_light}勒克斯"
    return (
        f"电量{battery}%，亮度{sensor.brightness}%，音量{sensor.volume}%，"
        f"步数{sensor.steps}，{motion}，{sensor.day_phase}{light}"
    )


def build_prompt(sensor: SensorPayload, *, nonce: int = 0) -> str:
    salt = nonce
    example = _FEW_SHOT_EXAMPLES[salt % len(_FEW_SHOT_EXAMPLES)]
    return (
        "写一条中文废话预言：第二人称「你」，32-42字，冷幽默荒诞，只输出预言正文。\n\n"
        f"可参考（不必写入正文）数据：{_sensor_facts(sensor)}\n"
        f"风格参考：{example}"
    )


def build_mlx_chat(sensor: SensorPayload, *, nonce: int = 0) -> MlxChatPrompt:
    salt = nonce
    example = _FEW_SHOT_EXAMPLES[salt % len(_FEW_SHOT_EXAMPLES)]
    user = (
        f"可参考数据：{_sensor_facts(sensor)}\n"
        f"风格参考：{example}"
    )
    return MlxChatPrompt(system=_SYSTEM_PROMPT, user=user)
